nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/nar/StandardNarManager.java

```diff
@@ -0,0 +1,504 @@
+                } else {
+                    LOGGER.info("Coordinator has NAR [{}] which exists locally with a matching digest, will not download", coordinatorNarId);
+                }
+            }
+        } catch (final Exception e) {
+            // if the current node has existing NARs and the sync fails then we throw an exception to fail start up, otherwise we don't know if the digests of the local NARs
+            // match with the coordinator which could result in joining the cluster and running slightly different code on one node
+            // if the current node has no existing NARs then we can let the node proceed and attempt to join the cluster because maybe the cluster coordinator had no NARs anyway,
+            // and if it did then flow synchronization will fail after this because the local flow will have ghosted components that are not ghosted in the cluster
+            if (localNarCountBeforeSync > 0) {
+                throw new RuntimeException("Failed to sync NARs from cluster coordinator", e);
+            } else {
+                LOGGER.error("Failed to sync NARs from cluster coordinator, no NARs exist locally, will proceed", e);
+            }
+        }
+    }
+
+    private void downloadNar(final NarRestApiClient narRestApiClient, final NarSummaryDTO narSummary) throws IOException {
+        try (final InputStream coordinatorNarInputStream = narRestApiClient.downloadNar(narSummary.getIdentifier())) {
+            final NarInstallRequest installRequest = NarInstallRequest.builder()
+                    .source(NarSource.valueOf(narSummary.getSourceType()))
+                    .sourceIdentifier(narSummary.getSourceIdentifier())
+                    .inputStream(coordinatorNarInputStream)
+                    .build();
+            installNar(installRequest, false);
+        }
+    }
+
+    private NarSummariesEntity getNarSummariesFromCoordinator(final NarRestApiClient narRestApiClient) {
+        final Instant waitUntilInstant = Instant.ofEpochMilli(System.currentTimeMillis() + MAX_WAIT_TIME_FOR_NARS.toMillis());
+        while (System.currentTimeMillis() < waitUntilInstant.toEpochMilli()) {
+            final NarSummariesEntity narSummaries = listNarSummaries(narRestApiClient);
+            if (narSummaries != null) {
+                return narSummaries;
+            }
+            LOGGER.info("Unable to retrieve NAR summaries from cluster coordinator, will retry until [{}]", waitUntilInstant);
+            sleep(Duration.ofSeconds(5));
+        }
+        return null;
+    }
+
+    private NarSummariesEntity listNarSummaries(final NarRestApiClient narRestApiClient) {
+        try {
+            return narRestApiClient.listNarSummaries();
+        } catch (final NarRestApiRetryableException e) {
+            LOGGER.warn("[{}], will retry", e.getMessage());
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("", e);
+            }
```

Comment:
The reason I did this was because it is normal that the listing will need to be retied a couple of times while possibly waiting for the coordinator to be ready to serve requests, so printing a huge stack trace each time makes it look like a major problem when it most likely isn't. I'll probably remove this since it is likely not needed.